lib/offsite_payments/integrations/paytm.rb

```diff
@@ -0,0 +1,298 @@
+module OffsitePayments #:nodoc:
+  module Integrations #:nodoc:
+    module Paytm
+      require 'openssl'
+      require 'base64'
+      require 'digest'
+      require 'securerandom'
+
+      mattr_accessor :test_url
+      mattr_accessor :production_url
+
+      self.test_url = 'https://pguat.paytm.com/oltp-web/processTransaction'
+      self.production_url = 'https://secure.paytm.in/oltp-web/processTransaction'
+
+      def self.service_url
+        # self.production_url
+        OffsitePayments.mode == :production ? production_url : test_url
+      end
+
+      def self.notification(post, options = {})
+        Notification.new(post, options)
+      end
+
+      def self.return(post, options = {})
+        Return.new(post, options)
+      end
+
+      def self.checksum(params, secret_key, salt = nil)
+        salt = SecureRandom.urlsafe_base64(4 * (3 / 4)) if salt.nil?
+        keys = params.keys
```

Comment:
I tried running the gateway with sandbox credentials but it failed at this step. `form_fields` method errors out saying: `undefined method 'keys' for #<Array:0x007fbae0c3e2c0>`. Your code is doing `Paytm.checksum(checksum_payload_items, @options[:credential2])` in a previous step and `checksum_payload_items` is actually an array, not a hash, so `params.keys` will fail.